File: python/examples/edgejs_http.py
import asyncio
import socket
import urllib.request
from pathlib import Path

from wasmer_sdk import Wasmer


SERVER_SOURCE = (
    Path(__file__).resolve().parents[2]
    / "fixtures/edgejs/server.js"
)


async def wait_for_line(lines, marker: str) -> None:
    async for line in lines:
        print(line)
        if marker in line:
            return
    raise RuntimeError(f"EdgeJS exited before emitting {marker!r}")


def fetch(port: int) -> tuple[int, str]:
    request = urllib.request.Request(
        f"http://127.0.0.1:{port}/hello",
        headers={"Connection": "close"},
    )
    with urllib.request.urlopen(request, timeout=5) as response:
        return response.status, response.read().decode()


def reserve_port() -> int:
    with socket.socket() as listener:
        listener.bind(("127.0.0.1", 0))
        return listener.getsockname()[1]


async def main() -> None:
    port = reserve_port()

    async with Wasmer(output_bytes=256 * 1024) as client:
        sandbox = await client.sandboxes.create(
            packages=["wasmer/edgejs-quickjs@0.1.0"],
            files={"server.js": SERVER_SOURCE.read_bytes()},
            env={"PORT": str(port)},
            network="host",
        )
        async with sandbox:
            process = await sandbox.command(
                "edge",
                ["/workspace/server.js"],
            ).spawn(
                stdout="pipe",
                stderr="capture",
                output_bytes=256 * 1024,
            )
            assert process.stdout is not None

            try:
                await asyncio.wait_for(
                    wait_for_line(
                        process.stdout.lines(),
                        "Edge.js listening on",
                    ),
                    timeout=20,
                )
                status, body = await asyncio.to_thread(fetch, port)
                print(f"GET /hello -> {status}")
                print(body)
            finally:
                await process.terminate(grace_period=2)
                await process.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: js/examples/edgejs_http.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import net from "node:net";

import { Wasmer } from "@wasmer/sdk/node";

const serverSource = await readFile(
  new URL("../../fixtures/edgejs/server.js", import.meta.url),
);
const port = await reservePort();
const wasmer = new Wasmer();
const sandbox = await wasmer.sandboxes.create({
  packages: ["wasmer/edgejs@0.2.0"],
  files: { "server.js": serverSource },
  env: { PORT: String(port) },
  network: { mode: "host" },
});
const guest = await sandbox
  .command("edge", ["/workspace/server.js"])
  .spawn({ stdout: "pipe", stderr: "capture" });

try {
  assert(guest.stdout);
  await waitForLine(guest.stdout.lines(), "Edge.js listening on", 20_000);

  const response = await fetch(`http://127.0.0.1:${port}/hello`, {
    headers: { connection: "close" },
  });
  console.log(`GET /hello -> ${response.status}`);
  console.log(await response.text());
} finally {
  await guest.terminate({ gracePeriodMs: 2_000 });
  await guest.wait();
  await sandbox.close();
  await wasmer.close();
}

async function waitForLine(lines, marker, timeoutMs) {
  let timeout;
  try {
    await Promise.race([
      (async () => {
        for await (const line of lines) {
          console.log(line);
          if (line.includes(marker)) return;
        }
        throw new Error(`process exited before emitting ${marker}`);
      })(),
      new Promise((_, reject) => {
        timeout = setTimeout(
          () => reject(new Error(`timed out waiting for ${marker}`)),
          timeoutMs,
        );
      }),
    ]);
  } finally {
    clearTimeout(timeout);
  }
}

async function reservePort() {
  const server = net.createServer();
  await new Promise((resolve, reject) => {
    server.once("error", reject);
    server.listen(0, "127.0.0.1", resolve);
  });
  const address = server.address();
  assert(address && typeof address !== "string");
  await new Promise((resolve, reject) => {
    server.close((error) => (error ? reject(error) : resolve()));
  });
  return address.port;
}
